syntax = "proto3";


package hyrl;

service ObstacleAvoidanceService {
  rpc GetDirection(DirectionRequest) returns (DirectionResponse);
  rpc GetTrajectory(TrajectoryRequest) returns (TrajectoryResponse);
}

// define drone state
message DroneState {
  float x = 1;
  float y = 2;
  float z = 3;
}

message TrajectoryRequest{
  DroneState current_state = 1;
  DroneState target_state = 2;
  uint32 num_waypoints = 3; 
  uint32 duration_s = 4; 
}


message TrajectoryResponse {
  repeated DroneState trajectory = 1;
}

// DEFINE DIRECTION REQUEST
message DirectionRequest {
  enum ModelType {
    STANDARD = 0;
    HYBRID = 1;
  }
  DroneState drone_state = 1;
  ModelType model_type = 2;
}

// DEFINE DIRECTION RESPONSE
message DirectionResponse {
  oneof control_type {
    DiscreteHeading discrete_heading = 2;
    ContinuousHeading continuous_heading = 3;
  }
}

message DiscreteHeading { 
  HeadingDirection direction = 1;
}

enum HeadingDirection {
  RESERVED = 0;
  STRAIGHT = 1;
  LEFT = 2;
  HARD_LEFT = 3;
  RIGHT = 4;
  HARD_RIGHT = 5;
}

message ContinuousHeading { float rad = 1; }
